Cython/Build/Tests/TestRecythonize.py

```diff
@@ -0,0 +1,213 @@
+import shutil
+import os
+import tempfile
+
+import Cython.Build.Dependencies
+import Cython.Utils
+from Cython.TestUtils import CythonTest
+
+
+
+SAME = "The result of cytonization is the same"
+INCORRECT = "Incorrect cythonization"
+LINE_BEFORE_IMPLEMENTATION = '  /* "{filename}":{at_line}\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
+
+
+class TestRecythonize(CythonTest):
+    language_level = 3
+    dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+
+    def сlear_function_and_Dependencies_caches(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
+
+    def setUp(self):
+        CythonTest.setUp(self)
+        self.сlear_function_and_Dependencies_caches()
+        self.temp_dir = (
+            tempfile.mkdtemp(
+                prefix='recythonize-test',
+                dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
+            )
+        )
+        self.src_dir = os.path.join(self.temp_dir, 'src')
+        os.mkdir(self.src_dir)  # self.temp_dir is already a temp
+
+    def tearDown(self):
+        CythonTest.tearDown(self)
+        self.сlear_function_and_Dependencies_caches()
+        shutil.rmtree(self.temp_dir)
+
+    def write_to_file(self, path, text):
+        with open(path, "w") as f:
+            f.write(text)
+
+    def fresh_cythonize(self, *args, **kwargs):
+        self.сlear_function_and_Dependencies_caches()
+        kwargs.update(language_level=self.language_level)
+        Cython.Build.Dependencies.cythonize(*args, **kwargs)
+
+    def refresh_dep_tree(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Utils.clear_method_caches(self.dep_tree)
+        self.dep_tree._transitive_cache.clear()
+
+    def fresh_all_dependencies(self, *args, **kwargs):
+        self.refresh_dep_tree()
+        return self.dep_tree.all_dependencies(*args, **kwargs)
+
+    def relative_lines(self, lines, line, start, end):
+        try:
+            ind = lines.index(line)
+            return lines[ind+start: ind+end]
+        except ValueError:
+            # XXX: It is assumed that VARS_LINE is always present.
+            ind = lines.index(VARS_LINE)
+            raise ValueError(
+                "{0!r} was not found, presumably in {1}".format(
+                    line, lines[ind-10: ind-1]))
```

Comment:
Should i replace `lines[ind-10: ind-1]` with `"".join(lines[ind-10: ind-1])`?

And therefore `{1}` to `{1!r}`

Reply:
Sounds better to have a string, but I haven't seen the output. Whatever you think is clearer when the test fails.

Reply:
join + {1}:
```python
ValueError: '  /* "a.pyx":1\n-' was not found, presumably in
  /* "a.pyx":1
 * x = 1             # <<<<<<<<<<<<<<
 */
  __pyx_v_1a_x = 1.0;
  __pyx_t_1 = __Pyx_PyDict_NewPresized(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_1) < 0) __PYX_ERR(0, 1, __pyx_L1_error)
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
```

join + {1!r}:
```python
ValueError: '  /* "a.pyx":1\n-' was not found, presumably in '\n  /* "a.pyx":1\n * x = 1             # <<<<<<<<<<<<<<\n */\n  __pyx_v_1a_x = 1.0;\n  __pyx_t_1 = __Pyx_PyDict_NewPresized(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1, __pyx_L1_error)\n  __Pyx_GOTREF(__pyx_t_1);\n  if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_1) < 0) __PYX_ERR(0, 1, __pyx_L1_error)\n  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;\n'   
```

no join + {1}:
```python
ValueError: '  /* "a.pyx":1\n-' was not found, presumably in ['\n', '  /* "a.pyx":1\n', ' * x = 1             # <<<<<<<<<<<<<<\n', ' */\n', '  __pyx_v_1a_x = 1.0;\n', '  __pyx_t_1 = __Pyx_PyDict_NewPresized(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 1, __pyx_L1_error)\n', '  __Pyx_GOTREF(__pyx_t_1);\n', '  if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_1) < 0) __PYX_ERR(0, 1, __pyx_L1_error)\n', '  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;\n']
```

Reply:
> Sounds better to have a string, but I haven't seen the output. Whatever you think is clearer when the test fails.

@scoder The point is that in the case of join + {1} we have a nice output, but in the case of no join + {1}/{1!r} it is easier to understand why the search failed, because we see where and how the lines end.
In the third case, we can try to understand what is wrong with the search, but this is a little more difficult, since the line is merged into one and we do not see where it ends immediately.

I think if we conclude in the form
```python
ValueError: 'line' was not found, presumably in
'line x'
'line x + 1'
```
it will be both nice and easy to look for an error.

Reply:
resolved [here](https://github.com/cython/cython/pull/4241/commits/bc1ad78558e36b8d360824d3cf983b7c7db998eb)

Reply:
> resolved [here](https://github.com/cython/cython/pull/4241/commits/bc1ad78558e36b8d360824d3cf983b7c7db998eb)

implemented this idea

> I think if we conclude in the form
> 
> ```python
> ValueError: 'line' was not found, presumably in
> 'line x'
> 'line x + 1'
> ```
> 
> it will be both nice and easy to look for an error.